src/ape/managers/chain.py

```diff
@@ -408,18 +414,82 @@ def outgoing(self) -> Iterator[ReceiptAPI]:
         """
         All outgoing transactions, from earliest to latest.
         """
-        explorer = self.provider.network.explorer
-        explorer_receipts = (
-            [r for r in explorer.get_account_transactions(self.address)] if explorer else []
-        )
-        for receipt in explorer_receipts:
-            if receipt.sender != self.address:
-                # Likely ``self.address`` is a contract.
-                # Cache the receipts by their sender instead and skip them here.
-                self.chain_manager.history.append(receipt)
-                continue
+
+        start_nonce = 0
+        stop_nonce = self.__len__() - 1  # just to cache this value
+
+        # TODO: Add ephemeral network sessional history to `ape-cache` instead,
+        #       and remove this (replace with `yield from iter(self[:len(self)])`)
+        for receipt in self.sessional:
+            if receipt.nonce < start_nonce:
+                raise QueryEngineError("Sessional history corrupted")
+
+            if receipt.nonce > start_nonce:
+                # NOTE: There's a gap in our sessional history, so fetch from query engine
+                yield from iter(self[start_nonce : receipt.nonce + 1])  # noqa: E203
 
             yield receipt
+            start_nonce = receipt.nonce + 1  # start next loop on the next item
+
+        if start_nonce != stop_nonce:
+            # NOTE: there is no more sessional history, so just return query engine iterator
+            yield from iter(self[start_nonce : stop_nonce + 1])  # noqa: E203
+
+    def query(
+        self,
+        *columns: List[str],
```

Comment:
It is a bit weird querying the transaction:

```
account.history.query("transaction", stop_nonce=4)
```

gives me:

```
                                         transaction
0  DynamicFeeTransaction:\n  chainId: 1\n  to: 0x...
1  DynamicFeeTransaction:\n  chainId: 1\n  to: 0x...
2  DynamicFeeTransaction:\n  chainId: 1\n  to: 0x...
3  DynamicFeeTransaction:\n  chainId: 1\n  to: 0x...
```

which is not super useful. Maybe we could make them JSON or something?

Reply:
Yeah, I think this has to do with the model we have for ReceiptAPI (although I think due to how it's getattr method works you can actually query properties of it directly)

Reply:
Resolved with [bd12a91](https://github.com/ApeWorX/ape/pull/1277/commits/bd12a9141244d57e5bbc8c3af2a451a42b24cc2c)

although it's a general performance improvement too